bsc.generics/GenericTests.bs

```diff
@@ -0,0 +1,175 @@
+package GenericTests where
+
+import ListN
+import Vector
+import BuildVector
+
+-- Represents "evidence" for type equality between a and b,
```

Comment:
This particular idiom was taken from dependently-typed languages, e.g. Idris: http://docs.idris-lang.org/en/latest/tutorial/theorems.html#equality
Bluespec doesn't have GADT type constructors, but we can sort of mimic this with a cleverly-typed function.

Using a type class is conceptually similar (using the instance dictionary object rather than an explicit data value), but as you said doesn't quite work without some way to actually demand the existence of the instance.  So in this particular case I don't think that using a type class would help to make this any more ergonomic.  

> That also means that the code in sysGenericTests isn't the entirety of the testing happening in this file. The typechecking of many other defs in this file are also additional tests (whose success is silence). That wasn't immediately obvious to me. 

I'm not sure that I quite follow.  Yes, we want to be sure that the derived Generic instances compile, but I'm not sure if this is what you are referring to.  Or do you mean the tests using this mechanism that follow the comment `-- Test generic representations`?

Reply:
My last comment was just to say that several independent tests coexist in this file.  Normally when a test has a `sysFoo` module in it, that is the only test in the file -- in this case, I scrolled to that module, read it, and then mistakenly thought that I'd understood everything being tested in this file.  I later realized that the definitions following the `-- Test generic representations` comment are independent tests that are also contained in this file.

This was merely a comment about organization.  I'm not proposing any changes (unless you feel you want to change anything based on this feedback).  I don't know that more comments would help, and it may not be worth separating out the other tests into their own file.